File: tests/stylex_validation_create_theme_test/stylex_validation_create_theme_test.rs
use std::collections::HashMap;

use stylex_swc_plugin::{
    shared::structures::{
        named_import_source::RuntimeInjection, plugin_pass::PluginPass,
        stylex_options::StyleXOptionsParams,
    },
    ModuleTransformVisitor,
};
use swc_core::ecma::{
    parser::{Syntax, TsConfig},
    transforms::testing::test,
};

#[test]
#[should_panic(expected = "stylex.create calls must be bound to a bare variable.")]
fn must_be_bound_to_a_variable() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            stylex.createTheme({__themeName__: 'x568ih9'}, {});
        "#,
        r#""#,
        false,
    )
}

#[test]
#[should_panic(expected = "stylex() should have 1 argument.")]
fn it_must_have_two_arguments_no_args() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme();
        "#,
        r#""#,
        false,
    )
}

#[test]
#[should_panic(expected = "stylex() should have 1 argument.")]
fn it_must_have_two_arguments_one_args() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme({});
        "#,
        r#""#,
        false,
    )
}

#[test]
#[should_panic(expected = "Only static values are allowed inside of a stylex.create() call.")]
fn it_must_have_two_arguments_fn_args() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme(genStyles(),{});
        "#,
        r#""#,
        false,
    )
}

#[test]
#[should_panic(expected = "Can only override variables theme created with stylex.defineVars().")]
fn it_must_have_two_arguments_empty_object_args() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme({},{});
        "#,
        r#""#,
        false,
    )
}

test!(
    Default::default(),
    |tr| {
        ModuleTransformVisitor::new_test_styles(
            tr.comments.clone(),
            PluginPass::default(),
            Option::None,
        )
    },
    it_must_have_two_arguments_valid,
    r#"
        import stylex from 'stylex';
        export const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
            {}
        );
    "#
);

#[test]
#[should_panic(expected = "Only static values are allowed inside of a stylex.create() call.")]
fn variable_keys_must_be_a_static_value() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme(
                {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
                {[labelColor]: 'red',}
            );
        "#,
        r#""#,
        false,
    )
}

test!(
    Default::default(),
    |tr| {
        ModuleTransformVisitor::new_test_styles(
            tr.comments.clone(),
            PluginPass::default(),
            Option::None,
        )
    },
    values_must_be_static_number_or_string_in_stylex_create_theme_v1,
    r#"
        import stylex from 'stylex';
        export const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', cornerRadius: 'var(--cornerRadiusHash)'},
            {cornerRadius: 5,}
        );
    "#
);

test!(
    Default::default(),
    |tr| {
        ModuleTransformVisitor::new_test_styles(
            tr.comments.clone(),
            PluginPass::default(),
            Option::None,
        )
    },
    values_must_be_static_number_or_string_in_stylex_create_theme_v2,
    r#"
        import stylex from 'stylex';
        export const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
            {labelColor: 'red',}
        );
    "#
);

#[test]
#[should_panic(expected = "Only static values are allowed inside of a stylex.create() call.")]
fn values_must_be_static_number_or_string_in_stylex_create_theme_var() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme(
                {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
                {labelColor: labelColor,}
            );
        "#,
        r#""#,
        false,
    )
}


#[test]
#[should_panic(expected = "Only static values are allowed inside of a stylex.create() call.")]
fn values_must_be_static_number_or_string_in_stylex_create_theme_fn() {
    swc_core::ecma::transforms::testing::test_transform(
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        }),
        |tr| {
            ModuleTransformVisitor::new_test_styles(
                tr.comments.clone(),
                PluginPass::default(),
                Option::None,
            )
        },
        r#"
            import stylex from 'stylex';
            const variables = stylex.createTheme(
                {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
                {labelColor: labelColor(),}
            );
        "#,
        r#""#,
        false,
    )
}

File: packages/test-parser/output/__tests__/babel-plugin/stylex-validation-create-theme-test.js
function transform(source: string, opts: any = {}) {
    return transformSync(source, {
        filename: opts.filename || 'TestTheme.stylex.js',
        parserOpts: {
            flow: 'all'
        },
        plugins: [
            [
                stylexPlugin,
                {
                    stylexSheetName: '<>',
                    unstable_moduleResolution: {
                        type: 'haste'
                    },
                    ...opts
                }
            ]
        ]
    });
}
describe('@stylexjs/babel-plugin', ()=>{
    describe('[validation] stylex.createTheme()', ()=>{
        test('must be bound to a variable', ()=>{
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          stylex.createTheme({__themeName__: 'x568ih9'}, {});
        `);
            }).toThrow(messages.UNBOUND_STYLEX_CALL_VALUE);
        });
        test('it must have two arguments', ()=>{
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme();
        `);
            }).toThrow(messages.ILLEGAL_ARGUMENT_LENGTH);
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme({});
        `);
            }).toThrow(messages.ILLEGAL_ARGUMENT_LENGTH);
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme(genStyles(), {});
        `);
            }).toThrow(messages.NON_STATIC_VALUE);
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme({}, {});
        `);
            }).toThrow('Can only override variables theme created with stylex.defineVars().');
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme({__themeName__: 'x568ih9'}, genStyles());
        `);
            }).toThrow(messages.NON_STATIC_VALUE);
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme({__themeName__: 'x568ih9'}, {});
        `);
            }).not.toThrow();
        });
        test('variable keys must be a static value', ()=>{
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
            {[labelColor]: 'red',});
        `);
            }).toThrow(messages.NON_STATIC_VALUE);
        });
        test('values must be static number or string in stylex.createTheme()', ()=>{
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', cornerRadius: 'var(--cornerRadiusHash)'},
            {cornerRadius: 5,}
          );
        `);
            }).not.toThrow();
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
            {labelColor: 'red',}
          );
        `);
            }).not.toThrow();
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
            {labelColor: labelColor,}
          );
        `);
            }).toThrow(messages.NON_STATIC_VALUE);
            expect(()=>{
                transform(`
          import stylex from 'stylex';
          const variables = stylex.createTheme(
            {__themeName__: 'x568ih9', labelColor: 'var(--labelColorHash)'},
            {labelColor: labelColor(),}
          );
        `);
            }).toThrow(messages.NON_STATIC_VALUE);
        });
    });
});
